### ISSUE TITLE
Need to specify object priorities as an alternative to send order

### ISSUE THREAD

**Issue by huitema** [kixelated/warp-draft#139]
The current draft proposes that the transmission of objects be controlled by a "send order" (a.k.a. delivery order) property. This is very questionable, because the complete ordering using a send order property is only one of the plausible solutions for managing transmission. Other solutions are possible, and may result in better management of multiple tracks. For example, allocating priorities to layers in a layered encoding scheme allows both ordering of objects within a track and relative ordering of objects between tracks, without requiring fine grained coordination between the publishers of different tracks. I think that this has to be resolved before adoption by the working group.

I have discussed this topic privately with Suhas and then with Luke. I understand why Luke was initially pushing for the "send order" approach: it feels simpler than trusting relays to apply the right logic based on multiple parameters. But it is essentially a compressed encoding of a set of different properties, such as relative sending order of tracks, layers, groups, and object order within a group, with questions such as:

* if several video tracks are competing, should the choice to keep just the 720p layer or also try the 1080p layers be coordinated among tracks?
* Within a track, should a new group take complete precedence over the previous group? Or should the relay finish sending the 720p layer of the old group before starting the 1080p layer of the new group?
* Within a track, should there be some kind of timeout between groups, such as stop sending anything from group 5 if it is more than 200ms behind group 6? Or is that 2 seconds? Or 0 ms? 

With just a send order property, it is indeed possible to create a single send order for a bundle of tracks, but that requires coordination between the emitters of the tracks. The typical policy will also require some gymnastic with numbers: object with higher group IDs should typically be sent before objects with lower group ID, but within a group objects with lower group IDs should be sent first. If we follow the "lower send order first" convention, we end up encoding the send order as:

layer priority | max_group_number - group_number | object_number_within_group
---- | ---- | ----

This kind of ordering requires allocating enough bits to each component, which means knowing in advance the maximum number of groups within a session and the maximum number of objects within a group. For real time applications, this is problematic, because the session time is not bounded. It also feels redundant, as we already encode the group-id and the object-id as properties of each object. Attaching a layer prority to the object feels simpler than trying to jam it into a single send order property. (We would also save 3 or 4 bytes per packet, which is nice even if not a big priority.)

When multiple tracks compete, specifying a send order across tracks is harder, if only because we cannot assume that group numbers are allocated consistently across multiple tracks. Typically, we would want something like:

* consistent priority marking across tracks in a bundle
* send highest priority objects first, but do round robin allocation between tracks at a given priority level
* use group number and object number to select the next object to transmit within a track.

The "round robin" part is quite hard to do with a single "send order". It might be possible if a single emitter sends all tracks within a bundle, but we cannot make that hypothesis for real time conferences, with multiple participants each contributing a couple of tracks. Yet we would want the "bundle" to group all tracks within the conference, so that for example participants can obtain all the tracks except their own through a single QUIC connection. In contrast, the "round robin within a priority level" is trivially implemented if a track is mapped to a set of streams, with one stream per group and priority layer. 

Bottom line: we need an option to encode a priority number per object -- we might also call it a layer number. That should an alternative to encoding a send order.

**Comment by suhasHere** [comment]
This is something that was discussed at IETF116 about prioritization schemes.  It would help to document sendOrder , object priorities based schemes and commentary that spells the pros/cons like above as part of Section 7 for draft-05.

This will give WG a single place to start looking at the options being discussed, follow and contribute as we develop these  further.

**Comment by hardie** [comment]
>
> Bottom line: we need an option to encode a priority number per object --
> we might also call it a layer number. That should an alternative to
> encoding a send order.
>

(personal hat on)

During the discussions in Japan, I thought I heard support for introducing
a metavariable here; essentially, a way for a track to carry the
information about the type of ordering used within the track and for the
track bundle to indicate how the overarching priority among tracks should
be handled.  The approach of using per track ordering types allows us to
adapt in the future if there are new types of media (e.g. AR/VR) with
fundamentally different types of priority.  A registry of those would allow
both relatively simple priorities and complex ones to coexist.

I confess I didn't actually understand how to construct a useful track
bundle priority scheme without a great deal of complexity.  There are some
simple versions (prioritizing audio over video, for example) that work with
just an ordered list of the tracks, but there are lots of others where
getting a useful answer starts to look like it requires a turing-complete
syntax. Introducing a second metavariable here might allow us to punt the
problem down the road a bit, by letting us do the simple ordered track type
now as an exemplar, but push the more complex version down the road.  I
worry if we do that, though, that we will start to see a very early split
in client capabilities which will dim the usefulness of the overall system.

(personal hat off)

(chair hat on)

In general, I don't think the document needs to solve this problem at this
point; it can describe the problem and list it as an item to be determined
by the working group during the course of the work.  For document adoption,
this doesn't need to be done.

(chair hat off)

**Comment by suhasHere** [comment]
@hardie just to confirm, having the options for prioritization schemes listed will be beneficial for adoption since it gives sufficient starting place to start thinking about challenges in supporting unified use-cases and why these options exists.

I personally believe this will help with driving inputs from either side of the aisle.

**Comment by suhasHere** [comment]
my proposal will be "either we should document both the alternatives or none" to avoid confusions for WG members reading and willing to contribute.

**Comment by fluffy** [comment]
I think there are some key things we agreed on in Japan.  

1) The issue of which packet to send next is largely the same for clients and relays and we should describe this in one place in the moq-transport draft client section and point at it the relay section along with any parts that were somehow different for relays

2) We should describe the process in a very algorithmic way 

3) all the information used by the algorithm needs to be available to the relay from the object information 

I don't think we should have the draft be a blank sheet of paper on this because it is hard to understand how anything works without a place holder for this. I think we should write up two proposals for this in the draft then mark it as TBD to choose one.  

(I'm actually hopeful that when they are both written down, I can convince people that the huitema proposal meets the needs best and we can all get behind that one but I am fine adopting the a draft with two proposal and no decision between the two. )

**Comment by fluffy** [comment]
As a note on max group size.... 

Surveillance cameras produce video for years on the same session at high frame rates and short GOP sequences. 

It is common for "video conferences"  between ops control rooms to run for many days. Teams do things like just stand up a permanent video call between two conferences rooms they use to work in. 

It might be possible to come up with some logic for group numbers wrapping so we did not need as many bits.

**Comment by huitema** [comment]
Just for info, I keep this little table relating the frequency of a signal to the number of bits required to encode it, and specifically the maximum duration of a QUIC connection:

Bits used in groups | Video (conf) | Video (surv) | Audio
-- | -- | -- | --
Period(sec) | 30 | 2 | 0.02
Frequency (Hz) | 0.033 | 0.5 | 50
Beats per minute | 2 | 30 | 3000
Beats per hour | 120 | 1800 | 180000
Beats per day | 2880 | 43200 | 4320000
Beats per year | 1051920 | 15778800 | 1577880000
bits for one  year | 20.005 | 23.911 | 30.555
Minutes for 6 bits | 32 | 2.1 | 0.021
Hours for 14 bits | 137 | 9 | 0.091
Years for 30 bits | 1021 | 68 | 0.680
Years for 62 bits | 4.4E+12 | 2.9E+11 | 2.9E+09

**Comment by suhasHere** [comment]
From my reading the discussion, I do agree with some hacks or provisions, we can solve the max group size problem. but I would like to take a step back and see are we trying to solve a problem which may cease to exist depending on the approach

**Comment by huitema** [comment]
I think we have some kind of rough consensus on the following algorithm

1) priorities or ordering only works across multiple tracks if these tracks are part of the same "bundle". That's actually pretty much the definition of the bundle concept: a set of tracks that have been composed so the nodes can send them in a coordinated way. 

2) within a bundle, we expect a coordinated handling of priorities or lawyers. In layered encoding, we always have a partial ordering of layers within a track (in the mathematical sense of partial ordering, i.e., a tree of dependencies, not a straight line). The partial ordering can be transformed into a complete ordering by imposing application preferences such as "frame rate first" or "definition first". Once we have a complete ordering of layers within a track, we can assume that coordinated emitters can do a coordinated assignment of priorities to layers, so that these priorities can be applied across multiple tracks.

3) Within a track, we have multiple groups. Within a group, we have multiple layers. Layers with "highest priority" should always be sent before layers with lower priority. The priority values are typically set to "lower priority number first", as in PRI-0 before PRI-1, PRI-2 before PRI-3, etc.

4) Within a track, a group and a layer, we assume newer objects have dependencies on older objects. The consensus here is a bit rough, because some coding schemes allow for "restart points", such as "PRI-2 frames only have dependency to other PRI-2 frames sent after the PRI-2 restart point". This will probably be debated further as we gain experience. For now, we can simply assume that objects can only have dependencies on "object in the same group sent before on the same layer or on high priority layers".

4) Within a track, we have a rough consensus that objects in the next group shall be sent before objects in the previous group. This is only a rough consensus, because there are situations when waiting for the previous group to complete results in better experience. But we could take that as a starting point for now.

5) We cannot assume that group numbers are allocated in a coordinated way across tracks. Given that, we cannot assume that group numbers can be used to set the relative priority of objects belonging to different tracks.

Our challenge is to define a scheduling algorithm that works with these constraints and produces the expect results in a deterministic way. The algorithm may combine assigning objects to streams, which guarantees that objects on the same stream will be received in order. The algorithm may assume assigning priorities or send orders to streams, which guarantees that higher priority streams will be sent before lower priority ones. The algorithm may say something about what to do with streams of same priority, i.e., use FIFO, or Round Robin, or maybe weighted round robin, depending on what the stack supports.

**Comment by suhasHere** [comment]
Thanks Christian, I wonder if we have rough consensus on point 1 above. IIUC, we decided to revisit bundle at IETF 116 because we didn't think that is always possible to ensure end to end or we need to work out specifics.

**Comment by kixelated** [comment]
Ooops, somehow I didn't see this issue until now.


> With just a send order property, it is indeed possible to create a single send order for a bundle of tracks, but that requires coordination between the emitters of the tracks.

I think that's true for any prioritization scheme. There has to be some entity that knows how to coordinate tracks from separate sources. Even a simple rule of audio > video is not true for all tracks, nor does a relay even know if a track is audio or video.

Maybe a track could choose from a registry of prioritization schemes, but this only works if the tracks from separate sources have have agreed to use the same scheme. And even then, a prioritization scheme that prioritizes newer group IDs only works when tracks have aligned group IDs, which again requires cooperation (ex. same group duration).


> 1. priorities or ordering only works across multiple tracks if these tracks are part of the same "bundle". That's actually pretty much the definition of the bundle concept: a set of tracks that have been composed so the nodes can send them in a coordinated way.

Yeah exactly. I just don't see how a relay can prioritize tracks unless some entity has coordinated them already.


> I confess I didn't actually understand how to construct a useful track
bundle priority scheme without a great deal of complexity.  There are some
simple versions (prioritizing audio over video, for example) that work with
just an ordered list of the tracks, but there are lots of others where
getting a useful answer starts to look like it requires a turing-complete
syntax. Introducing a second metavariable here might allow us to punt the
problem down the road a bit, by letting us do the simple ordered track type
now as an exemplar, but push the more complex version down the road.  I
worry if we do that, though, that we will start to see a very early split
in client capabilities which will dim the usefulness of the overall system.

That's my thought process too. You start headed down this slippery slope, where you either need a registry of prioritization schemes for every use-case, or you need a turing-complete templating scheme. The more complicated we make it, the more "correct" the prioritization scheme becomes.

Ultimately though, relay support is the highest priority for any prioritization scheme (pun intended). Anything even remotely complicated will NOT be adopted by the wider ecosystem (ex. HTTP/2 priorities). I think being "correct" in every use-case is never going to be feasible.

My solution with send order is to punt the scheming to the entity to creates a bundle, so it's dirt simple for a relay to support. I think there's room in the middle, but I struggle to figure out what that looks like.

> But if we do adopt with the proposed "send order" text, we enshrine a proposal that does not actually work for the "real time conference" scenarios. If we want fast adoption, then we should leave the whole point of scheduling and ordering as TBD, and remove from the draft the text about "send order".

I don't understand why send order doesn't work for real-time conferencing scenarios.

The formulas you've posted make sense as possible prioritization schemes. When variables are known at encode time (ex. layer, group, object, timestamp, etc), then the send order can be computed by the encoder and written on the wire. If there's a variable that's NOT known at encode time (ex. current wall clock timestamp, queued object count) then we need a formula computed by each relay, but I'm skeptical if those are necessary.

It's fair to say that send order makes it difficult for tracks to be muxed, although that's not a problem specific to real-time conferencing.

**Comment by kixelated** [comment]
I thought maybe there's a world where we use send order within a track, and scale/offset that send order between tracks. 
You could write this formula on the wire so you could mux tracks by adding a new `case` statement. 

For example:
```
stream.send_order = switch object.track {
  case video: object.send_order / 60
  case audio: 44100 * object.send_order / 1024 + 3s
  case ad: object.send_order / 30 + 6432s
}
```

But the muxer still needs very specific information about how send order is computed for each track so it can coordinate them. For example, if:

```
video.send_order = object.group << 24 + object.layer << 16 - object.sequence
```

Ugh then the switch statement above is wrong, because the audio send order needs to be left shifted by 24. The muxer needs to know this formula to interleave send orders, and even then it's extremely difficult to perform this interleaving in the first place.


So I want to rule out muxing uncoordinated tracks. We can only interleave tracks that use compatible prioritization schemes (ie. send order formulas). This is a bundle, and between bundles, we only have the ability to use absolute (video > audio) or relative prioritization (audio = 2x video).

**Comment by kixelated** [comment]
And frankly, maybe we only need absolute and relative priorities between tracks. I thought I was clever by interleaving the send order between tracks, but I'm unconvinced that it actually makes a significant difference.

For example, I was using this at Twitch:
```
audio.send_order = timestamp - 3s
video.send_order = timestamp
```

The idea was audio is generally more important than video... unless you went through a tunnel for 10s. In which case I would rather deliver the latest video frame instead of audio from more than 3s ago.

But I was never able to validate this assumption, and frankly it's such an edge case that it barely matters anyway. Absolute prioritization between audio and video would work just fine. In fact relative prioritization between audio and video (even round-robin) would probably work too, since audio is smaller than video.

What about a priority per track, and send order within each track?

**Comment by huitema** [comment]
I was assuming absolute priorities, as in "audio track has absolute priority P1, 15fps video has P2, 30fps video has P3". If multiple tracks assign the same priorities the same way, then the corresponding streams are processed in parallel, round-robin way. That could work with user choices as well, as in "streams in track A have priority P1+X, P2+X, P3+X", but I really wonder whether we need that kind of complexity.

The problem is indeed transitions between groups, which are kind of internal to the track. At some point the relays will have to shed load, which may imply "reset the streams for the old group if they are not all sent already". So maybe P1 and P2 are all sent, but P3 is lagging, so some of P3 gets dropped.

This works well if there are multiple tracks with the same scheme. They do not have the same group boundaries, so the decision to reset some old track happens at different moments, but the result is still "P1 and P2 are all sent, but P3 is lagging, so some of P3 gets dropped for all tracks".

**Comment by kixelated** [comment]
> I was assuming absolute priorities, as in "audio track has absolute priority P1, 15fps video has P2, 30fps video has P3". If multiple tracks assign the same priorities the same way, then the corresponding streams are processed in parallel, round-robin way. That could work with user choices as well, as in "streams in track A have priority P1+X, P2+X, P3+X", but I really wonder whether we need that kind of complexity.
> 
> The problem is indeed transitions between groups, which are kind of internal to the track. At some point the relays will have to shed load, which may imply "reset the streams for the old group if they are not all sent already". So maybe P1 and P2 are all sent, but P3 is lagging, so some of P3 gets dropped.
> 
> This works well if there are multiple tracks with the same scheme. They do not have the same group boundaries, so the decision to reset some old track happens at different moments, but the result is still "P1 and P2 are all sent, but P3 is lagging, so some of P3 gets dropped for all tracks".

Yeah, and you still need the tracks to coordinate with these sorts of prioritization scheme. A track that purposely decides to use P3 will find itself losing bandwidth to a track that does not. This might not be intentional, such as an IPPPPP track accidentally being higher priority than an IBPBPB track because it does not support priority layers.

But a track could be higher priority on purpose. I'm thinking the main broadcast track in an NFL broadcast, which would always get bandwidth over the sideline or reaction track. These tracks would have to be intentionally encoded use P4+, aware that to the possibility that they could be bundled with the more important broadcast track at some future point.

I'm coming around to the belief that tracks from multiple sources being interleaved together is just a recipe for disaster. There's too many assumptions and moving pieces, and the side-effects won't be obvious due to the transient nature of congestion. A think only a single source (or explicitly coordinated sources) can make the sort of prioritization decision required to result in a positive outcome.

**Comment by hardie** [comment]
On Wed, Apr 26, 2023 at 1:48 AM Christian Huitema ***@***.***>
wrote:

> I was assuming absolute priorities, as in "audio track has absolute
> priority P1, 15fps video has P2, 30fps video has P3". If multiple tracks
> assign the same priorities the same way, then the corresponding streams are
> processed in parallel, round-robin way. That could work with user choices
> as well, as in "streams in track A have priority P1+X, P2+X, P3+X", but I
> really wonder whether we need that kind of complexity.
>
> (Individual hat on)

"if multiple tracks assign the same priorities the same way" is the key
statement and the corresponding question is how do you ensure that they
do?  There are three choices that are relatively common:  there is always a
single entity per user experience  assigning them and the entity guarantees
this ;  there is a single standardized way to make the assignment; or the
method of assignment is chosen from a known set and explicit (e.g. an IANA
registry name is in the metadata for the overall set of tracks.)  The other
method we have discussed (templating) makes explicit the different
assignment mechanisms without registries and with more degrees of freedom
but at the cost of a lot of complexity.

My experience with q-values leads me to believe that people end up treating
purely ordinal values as if the intervals among the values were fixed,
which is not a guaranteed property of an ordinal list (I got three bids for
my prized sow:  1000, 800, 300.  The order is clear, but the intervals are
not the same).  I worry that something similar could happen here for
priorities assigned by different parties but with different implicit
meanings.  I think that means we always want an explicit meaning for the
priority scheme.

To put this another way:  every party to the exchange should have the same
understanding of how the queues will be serviced based on the priority
values assigned.  I personally think that means a small number of
registered schemes, always explicitly called out.  That lets different
parties assign the priorities with the same understanding (Source A and
Source B only have to agree on Priority Scheme, not on a more complex
coordination), it lets the protocol evolve to use different priority
schemes for different types of media and/or user experiences, and it lets
you see instantly if there is a complex case (Audio tracks from source A
use the registered send-order priority while the timed text uses a
registered scheme that decays based on clock time).

I don't think we can simplify the use cases to always have a single entity
assigning the priorities for all tracks, and I am pretty sure we can pick a
single priority method that satisfies everyone.

(reminder, individual hat)

**Comment by huitema** [comment]
Well, we have a classic issue there:

1) Coordination between tracks is possible, but only between tracks that adhere to the same prioritization logic. 

2) If tracks pretend to be adhering to the logic but don't, the results are unpredictable.

3) If there are bugs in the logic chosen by the coordinated tracks or its implementation by emitters, the results are unpredictable.

4) If tracks do adhere to the same logic but the network of relays don't believe them and use their own logic, the results are also unpredictable.

5) if relays mix track that are not coordinated as if they were, the results are indeed unpredictable.

It seems to me that the solution has two components:

* there is some property of each track stating which prioritization scheme it is using
* there is an authority saying, "for the purpose of this session, track A, B and C can be mixed using prioritization scheme X"

**Comment by huitema** [comment]
Has of now, it is pretty clear that we won't quickly get consensus on this subject, and that waiting for that would delay draft adoption too much. I entered PR #149 to state that, and leave the debate unsolved for now.

**Comment by suhasHere** [comment]
> As of now, it is pretty clear that we won't quickly get consensus on this subject, and that waiting for that would delay draft adoption too much. I entered PR #149 to state that, and leave the debate unsolved for now.

#149  is a good starting point on this topic. I think we should get it merged to slow down the adoption

**Comment by afrind** [comment]
> I think we should get it merged to slow down the adoption

@suhasHere : to clarify : did you mean *speed up* the adoption, or you are asking to spend more time on this issue before adoption?

**Comment by afrind** [comment]
@huitema : are you OK with closing this now that the DISCUSS has merged?

**Comment by afrind** [comment]
Both proposals are in the draft.
